### ISSUE TITLE
GCP Instance templates / Instance group manager - error deleting

### ISSUE THREAD

**Issue by stakauskas** [pulumi/pulumi-gcp#680]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

When working with GCP and instance templates / instance groups, any failure to the run, corrupts the state without ability to cleanly recover. Say the casual scenario, where you create instance template and use instance group manager in GCP(along with other things). When provisioning, if something else fails during the run, Pulumi marks in the state that instance group should have new template(which was assigned during that run) and then on the next run tries to delete the previous template, but this actually being used as the previous update failed, so it cannot delete it. And this becomes real problem when having multiple instance groups. The only way to recover from it, is to create new templates, switch instance groups to use them and re-run Pulumi, so that it will be able to remove those marked for deletion(refresh does not help in this scenario). Example:

```
my-app-instance-template (mystack:group$mystack:mygroup$gcp:compute/instanceTemplate:InstanceTemplate)
error: deleting urn:pulumi:myproject::mystack::mystack:group$mystack:mygroup$gcp:compute/instanceTemplate:InstanceTemplate::my-app-instance-template: 1 error occurred:
	* Error waiting for Deleting Instance Template: The instance_template resource 'projects/myproject/global/instanceTemplates/my-app-instance-template-COMMIT_HASH' is already being used by 'projects/myproject/zones/myzone/instanceGroupManagers/my-app'
```

### Steps to reproduce

1. Create multiple compute instance templates and instance groups in GCP, setting instance template name to, say: `my-app-instance-template-COMMIT_HASH`
2. After provisioning, run Pulumi again with setting instance template to `my-app-instance-template-COMMIT_HASH2` and kill the deployment (process kill, network or anything else) in the middle of it, say when it starts creating instance templates or updating instance groups. Build will be marked as failed.
3. Run `pulumi refresh/up` again 
4. If the build was cancelled at the "right" point in time at step 2, the new `up` command will fail with the above message, that it cannot delete `my-app-instance-template-COMMIT_HASH`, but it cannot delete because actually the instance group is still using that template(because of failed run), yet Pulumi thinks, that it should be already using `my-app-instance-template-COMMIT_HASH2`.

Expected: not mark the template for deletion, unless the instance group manager has been actually assigned that template and previous instance was removed from the group.
Actual: Pulumi marks that instance group is running on `my-app-instance-template-COMMIT_HASH2`, while it actually running on `my-app-instance-template-COMMIT_HASH` and tries to remove `my-app-instance-template-COMMIT_HASH` unsuccessfully, as it's still being used by instance group manager.

**Comment by gorilych** [comment]
I've seen this this issue multiple times, but had not enough details for report. Thanks @stakauskas!

**Comment by jdoss** [comment]
I ran into this issue today when I was scaling down my instances to zero across three managed groups. 

```
Type                             Name                        Status                   Info
     pulumi:pulumi:Stack              internal-testing            **failed**               1 error
 +-  ├─ gcp:compute:InstanceTemplate  instancetemplate-command    **replacing failed**     1 error
 +-  ├─ gcp:compute:InstanceTemplate  instancetemplate-compute    **replacing failed**     1 error
 +-  └─ gcp:compute:InstanceTemplate  instancetemplate-ephemeral  **replacing failed**     1 error
 
Diagnostics:
  gcp:compute:InstanceTemplate (instancetemplate-compute):
    error: deleting urn:pulumi:testing::internal::gcp:compute/instanceTemplate:InstanceTemplate::instancetemplate-compute: 1 error occurred:
      * Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/instancetemplate-compute-dc279c0' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/mig-compute-98fe1f2'
 
  gcp:compute:InstanceTemplate (instancetemplate-ephemeral):
    error: deleting urn:pulumi:testing::internal::gcp:compute/instanceTemplate:InstanceTemplate::instancetemplate-ephemeral: 1 error occurred:
      * Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/instancetemplate-ephemeral-5fd9833' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/mig-ephemeral-e21d83a'
 
  gcp:compute:InstanceTemplate (instancetemplate-command):
    error: deleting urn:pulumi:testing::internal::gcp:compute/instanceTemplate:InstanceTemplate::instancetemplate-command: 1 error occurred:
      * Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/instancetemplate-command-340eef8' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/mig-command-f9e2c73'
```

Any chance to get some eyes on this from Pulumi staff? This is the number 1 thumbs up'd bug with the next closest issue having 5 thumbs.  

Not being able to reliably use instance templates and groups is a show stopper on adopting Pulumi for advanced deployments on GCP.

**Comment by stakauskas** [comment]
@stack72 - I see this has been closed, but the issue still very much active. Tested with most recent Pulumi versions:
```
        "@pulumi/gcp": "^6.30.0",
        "@pulumi/pulumi": "^3.35.3"
```

**Comment by jdoss** [comment]
Hey @stack72 I am just getting back into trying to use Instance Templates and I am seeing the error 


```
Diagnostics:
  gcp:compute:InstanceTemplate (instancetemplate-nomad-compute-test):
    error: deleting urn:pulumi:production::internal::gcp:compute/instanceTemplate:InstanceTemplate::instancetemplate-nomad-compute-test: 1 error occurred:
    	* Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/instancetemplate-nomad-compute-test-bd001d8' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/mig-nomad-compute-test-512b433'
 
  pulumi:pulumi:Stack (internal-production):
    error: update failed
```

I am using `v6.32.0` as https://github.com/pulumi/pulumi-gcp/issues/915 is blocking me from upgrading to the most recent version.

```
$ pulumi about
CLI          
Version      3.40.2
Go Version   go1.19.1
Go Compiler  gc

Plugins
NAME    VERSION
gcp     6.32.0
python  unknown
random  4.4.2

Host     
OS       fedora
Version  37
Arch     x86_64
```

**Comment by jdoss** [comment]
Trying things out with `name_prefix` still results in the same failure:

```
instance_template = gcp.compute.InstanceTemplate(
        f"instance-template-{instance_group}",
        name_prefix=f"{instance_group}-",
        machine_type=server_data["server_type"],
        tags=server_data["server_tags"],
        network_interfaces=[
        *snip*
```

```
Diagnostics:
  pulumi:pulumi:Stack (internal-production):
    error: update failed
 
  gcp:compute:InstanceTemplate (instance-template-nomad-compute-test):
    error: deleting urn:pulumi:production::internal::gcp:compute/instanceTemplate:InstanceTemplate::instance-template-nomad-compute-test: 1 error occurred:
    	* Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/nomad-compute-test-20220929224129348200000001' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/rigm-nomad-compute-test-b25d286'
```

**Comment by andygeorge** [comment]
+1, I am also seeing the same error when attempting deletions.

**Comment by jdoss** [comment]
@stack72 I tried a bunch of things to see if I could get the createBeforeDelete behavior so the new instance template is created and added to the regional instance group manager before the old one is removed and the only time I saw it work correctly was when I removed the `name_prefix=` from my example above. 

This caused Pulumi the resource to create the new template, and it added it to the regional instance group manager and then removed the old template. I was excited to think maybe not having any `name=` or `name_prefix=` on `gcp.compute.InstanceTemplate` would work around this issue but on a subsequent rerun of updating the instance template resulted in the same buggy behavior. ☹️ 

Maybe this is specific to using Python?

**Comment by stakauskas** [comment]
No, it's not Python only, and yes the bug is present on name and namePrefix. Even if you have namePrefix and stop Pulumi somewhere in the middle and you have say 20 templates to replace and it has marked them for replacement and Pulumi stopped (any kind of failure), on rerun it will fail with that error as it will try to delete current used templates.

The logic Pulumi manages state in these scenarios is completely wrong. In TF you can stop it any time, rerun and it will be fine. Even though this has been marked as closed, the bug is still present. Really shame that @stack72 closed it for no good reason.

**Comment by stakauskas** [comment]
@emiliza / @stack72 - can we re-open this please?

**Comment by jdoss** [comment]
This is still very much broken in pulumi-gcp 6.45.0.

```
Diagnostics:
  pulumi:pulumi:Stack (internal-production):
    error: update failed

  gcp:compute:InstanceTemplate (instance-template-nomad-compute):
    error: deleting urn:pulumi:production::internal::gcp:compute/instanceTemplate:InstanceTemplate::instance-template-nomad-compute: 1 error occurred:
      * Error waiting for Deleting Instance Template: The instance_template resource 'projects/mycoolproj/global/instanceTemplates/nomad-compute-xxx' is already being used by 'projects/mycoolproj/regions/us-central1/instanceGroupManagers/rigm-nomad-compute-xxx'
```

**Comment by jdoss** [comment]
@lblackstone @lukehoban can this get reopened?

**Comment by joeduffy** [comment]
@mikhailshilkov Re-opened - seems this is still an issue. Removing P0 for now until you can take a look.

**Comment by danielrbradley** [comment]
Appologies the follow-up responses got missed. I've added this to the priorities to be re-investigated after the new year.

**Comment by jdoss** [comment]
Hey @danielrbradley any word on when this can be fixed?

**Comment by andygeorge** [comment]
Hoping for an update as well, this is a blocker for using Pulumi over raw Terraform.

**Comment by lblackstone** [comment]
I started investigating this issue today, and have a consistent repro with the latest provider version:

Step 1: Create a stack with the following program
```ts
import * as gcp from "@pulumi/gcp";
import * as pulumi from "@pulumi/pulumi";

const myImage = pulumi.output(gcp.compute.getImage({
    family: "debian-11",
    project: "debian-cloud",
}));
const instanceTemplate = new gcp.compute.InstanceTemplate("it", {
    name: "test1",
    machineType: "e2-medium",
    region: "us-central1",
    networkInterfaces: [{
        network: "default",
    }],
    disks: [{
        sourceImage: myImage.selfLink,
    }],
}, {deleteBeforeReplace: true});

const appserver = new gcp.compute.InstanceGroupManager("appserver", {
    baseInstanceName: "app",
    zone: "us-central1-a",
    versions: [{
        instanceTemplate: instanceTemplate.id,
    }],
    targetSize: 2,
    namedPorts: [{
        name: "customhttp",
        port: 8888,
    }],
});
```

Step 2: Change the name of the `InstanceTemplate`
Step 3: Run another `pulumi up`. The update will fail:
```
pulumi up
Previewing update (dev)

View Live: https://app.pulumi.com/lblackstone/gcp-test/dev/previews/3ec0535c-0fcd-4c10-9a81-245a2145a8be

     Type                                 Name          Plan        Info
     pulumi:pulumi:Stack                  gcp-test-dev
 +-  ├─ gcp:compute:InstanceTemplate      it            replace     [diff: ~name]
 ~   └─ gcp:compute:InstanceGroupManager  appserver     update      [diff: ~versions]


Resources:
    ~ 1 to update
    +-1 to replace
    2 changes. 1 unchanged

Do you want to perform this update? yes
Updating (dev)

View Live: https://app.pulumi.com/lblackstone/gcp-test/dev/updates/18

     Type                             Name          Status                   Info
     pulumi:pulumi:Stack              gcp-test-dev  **failed**               1 error
 +-  └─ gcp:compute:InstanceTemplate  it            **replacing failed**     1 error


Diagnostics:
  pulumi:pulumi:Stack (gcp-test-dev):
    error: update failed

  gcp:compute:InstanceTemplate (it):
    error: deleting urn:pulumi:dev::gcp-test::gcp:compute/instanceTemplate:InstanceTemplate::it: 1 error occurred:
    	* Error waiting for Deleting Instance Template: The instance_template resource 'projects/pulumi-development/global/instanceTemplates/test1' is already being used by 'projects/pulumi-development/zones/us-central1-a/instanceGroupManagers/appserver-b34c08d'

Resources:
    1 unchanged

Duration: 13s
```

```
pulumi about
CLI
Version      3.51.1
Go Version   go1.19.5
Go Compiler  gc

Plugins
NAME    VERSION
gcp     6.47.0
nodejs  unknown

Host
OS       darwin
Version  12.6.1
Arch     arm64

This project is written in nodejs: executable='/Users/levi/.nvm/versions/node/v16.10.0/bin/node' version='v16.10.0'

Current Stack: lblackstone/gcp-test/dev

TYPE                                                   URN
pulumi:pulumi:Stack                                    urn:pulumi:dev::gcp-test::pulumi:pulumi:Stack::gcp-test-dev
pulumi:providers:gcp                                   urn:pulumi:dev::gcp-test::pulumi:providers:gcp::default_6_47_0
gcp:compute/instanceTemplate:InstanceTemplate          urn:pulumi:dev::gcp-test::gcp:compute/instanceTemplate:InstanceTemplate::it
gcp:compute/instanceGroupManager:InstanceGroupManager  urn:pulumi:dev::gcp-test::gcp:compute/instanceGroupManager:InstanceGroupManager::appserver


Found no pending operations associated with dev

Dependencies:
NAME            VERSION
@pulumi/gcp     6.47.0
@pulumi/pulumi  3.52.0
@types/node     14.18.36

Pulumi locates its logs in /var/folders/ny/f_y5fsqd235fpx5bs6ghyk4w0000gn/T/ by default
```

**Comment by lblackstone** [comment]
I haven't been able to reproduce the issue with the OP's steps so far. This seems to be a problem with the operation ordering, where Pulumi tries to delete the existing `InstanceTemplate` (deleteBeforeReplace) prior to creating the replacement. I did notice that cancelling the update while a replacement `InstanceTemplate` was creating can leak a resource that isn't tracked by Pulumi, so it's possible that stopping the update at just the right moment leads to the out of order condition.

I'm still unclear on the following points:

1. Does this issue only occur if an update was cancelled?
2. Has the behavior changed since the original report?
3. Does running a `refresh` affect the behavior?
4. Are multiple `InstanceTemplate`/`InstanceGroupManager` resources needed, or is this reproducible with one pair?

I'll keep digging on this, but if anyone has more information about reproducing the issue reliably, or clarifying the above points, that would be helpful.

**Comment by jdoss** [comment]
This is my experience with the issue. 

1. Instance Template updates (user data changes)
2. Pulumi tries to replace the template but it cannot be cause it try's to delete old one while it is still in use by the Instance Group. 
``` gcp:compute:InstanceTemplate (instancetemplate-nomad-compute-test):
    error: deleting urn:pulumi:production::internal::gcp:compute/instanceTemplate:InstanceTemplate::instancetemplate-nomad-compute-test: 1 error occurred:
    	* Error waiting for Deleting Instance Template: The instance_template resource 'projects/eng-internal/global/instanceTemplates/instancetemplate-nomad-compute-test-bd001d8' is already being used by 'projects/eng-internal/regions/us-central1/instanceGroupManagers/mig-nomad-compute-test-512b433'
``` 

You can work around this by creating a junk template and then you manually switch the instance group to use that junk template and then run `pulumi up` and it can delete the old template and replace the junk template on the instance group with the updated new one.

For some reason Pulumi is not creating the new instance template, attaching the new one to the instance group, and then deleting the old one. It is trying to delete the old one while it is still attached to the instance group and that causes the error.

**Comment by lblackstone** [comment]
Thanks for the update. I think I've finally tracked down a repro that matches the behavior you're seeing.

1. Create `InstanceTemplate` + `InstanceGroupManager`
2. Change a property of the `InstanceTemplate` to cause a replacement
3. Cancel the update as soon as the `InstanceGroupManager` update operation starts
4. Change a property of the `InstanceTemplate` to cause a replacement
5. Run another update and the replacement will fail because the old `InstanceTemplate` is still in use

Here's the program I used:

```ts
import * as gcp from "@pulumi/gcp";
import * as pulumi from "@pulumi/pulumi";

const myImage = pulumi.output(gcp.compute.getImage({
    family: "debian-11",
    project: "debian-cloud",
}));
const instanceTemplate = new gcp.compute.InstanceTemplate("it", {
    machineType: "e2-small",
    region: "us-central1",
    networkInterfaces: [{
        network: "default",
    }],
    disks: [{
        sourceImage: myImage.selfLink,
    }],
    tags: ["a"], // Change to "b" in step 2
});

const appserver = new gcp.compute.InstanceGroupManager("appserver", {
    baseInstanceName: "app",
    zone: "us-central1-a",
    versions: [{
        name: "appserver",
        instanceTemplate: instanceTemplate.id,
    }],
    targetSize: 0,
    namedPorts: [{
        name: "customhttp",
        port: 8888,
    }],
});
```

This definitely seems like a bug in the update logic if the operation was cancelled before the `InstanceGroupManager` completes.

**Comment by jdoss** [comment]
I am confused. My run fails without cancelling anything. It fails as soon as Pulumi tries to delete the old `InstanceTemplate` that is still in use by the `InstanceGroupManager`. It needs to:

1. Create the new `InstanceTemplate`
2. Remove the old `InstanceTemplate` off of the `InstanceGroupManager`
3. Add the new `InstanceTemplate` to the `InstanceGroupManager`
4. Delete the old `InstanceTemplate`

**Comment by lblackstone** [comment]
@jdoss Have you tried running a refresh as part of the update? The OP's problem involved cancellation, but it's possible that something else is getting you into a similar state. If so, the problem is that Pulumi's state doesn't match the GCP state, which results in an incorrect order of operations. Running a refresh ensures that the state is accurate before the update.

**Comment by jdoss** [comment]
Yeah I have refreshed before hand and it still breaks when it tries to delete the `InstanceTemplate` when it is still in use on the `InstanceGroupManager`. 

Is there a way to pass in 

```
lifecycle {
    create_before_destroy = true
  }
```

Like you can in Terraform? https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance_template#using-with-instance-group-manager

Pulumi should be creating the new template, replacing the old one on the `InstanceGroupManager` and then deleting the old one.

**Comment by lblackstone** [comment]
@jdoss Unlike Terraform, Pulumi uses "create before destroy" behavior by default for replacements. The [deleteBeforeReplace option](https://www.pulumi.com/docs/intro/concepts/resources/options/deletebeforereplace/) can be used to invert that behavior.

I think we're going to need more information about your setup to continue troubleshooting. After extensive testing, refresh has always fixed it for me when the provider is in this state. My guess is that either you're hitting a separate issue with similar symptoms, or something about your configuration is forcing the `deleteBeforeReplace` behavior.

Can you provide the code that you're using to configure the `InstanceTemplate` and `InstanceGroupManager` resources? It would also be helpful to get [verbose logs](https://www.pulumi.com/docs/support/troubleshooting/#verbose-logging) from a stack update where you're seeing this problem. You can email them or DM on community slack. Thanks!

**Comment by lblackstone** [comment]
I followed up with @jdoss offline, and we figured out the problem. One of the resources (a `BucketObject`) that the `InstanceTemplate` depends on had a name assigned rather than using [auto-naming](https://www.pulumi.com/docs/intro/concepts/resources/names/). When the `BucketObject` was updated, it had to be `deleteBeforeReplace`d, which also caused a `deleteBeforeReplace` on the `InstanceTemplate`. When we changed the `BucketObject` to use auto-naming, the update ordering worked as expected.

In summary, this issue occurs whenever an `InstanceGroupManager`/`RegionInstanceGroupManager` is pointing to an `InstanceTemplate` that gets `deleteBeforeReplace`d. This can be avoided by:

1. Refreshing the stack if a previous update was interrupted, which [can cause ordering problems](https://github.com/pulumi/pulumi-gcp/issues/680#issuecomment-1401114976)
2. Make sure that any resources that feed into `InstanceTemplate` resources are not being `deleteBeforeReplace`d. This can occur if the resource has a name that is manually specified rather than using auto-naming.

I'm going to close this issue out as resolved, and thanks to everyone for their patience while we tracked this down!
